README.md
# Minirandr

Minirandr is a very simple command line interface to `xrandr` for configuring screens when all you care about is which screens to connect, and how they're arranged. All screens will be connected at native resolution in standard orientation.

## Quick start

1. Grab a binary from [Github Releases](https://github.com/safehammad/minirandr/releases) and place it anywhere on your `PATH`.

2. Run `minirandr` to list connected screens. For example:
```
    $ minirandr
    0: eDP1 1920x1080
    1: DP1 1920x1200
    2: HDMI2 2560x1080
```

3. Configure your screens left to right. For example, to configure screen 1 on the left, screen 0 to the right of that as primary screen, and screen 2 to the right of that, run:

    `$ minirandr 1 0p 2`

## Installation

This software relies on the [xrandr](https://www.x.org/releases/X11R7.5/doc/man/man1/xrandr.1.html) command to configure screens. The command is installed by default on most Linux distros, and can be installed on Mac (though I've not tested this). I'm not aware of a Windows port of `xrandr`.

https://github.com/babashka/babashka/releases

### Option 1: Download binary

Grab a binary from [Github Releases](https://github.com/safehammad/minirandr/releases) and place it anywhere on your `PATH`.

### Option 2: Run as Babashka script

This application is written as a Clojure script which can be run from the shell using [Babashka](https://babashka.org/). Install instructions for Babashka can be found at https://github.com/babashka/babashka#installation.

Then simply run the provided shell script `./bin/minirandr`. For convenience, you can add the `bin` directory to your `PATH`.

### Option 3: Build native Executable

You can build a native executable for your own platform using [GraalVM](https://www.graalvm.org/) as follows:

1. Install Clojure -- the [Getting Started Guide](https://clojure.org/guides/getting_started) will get you up and running quickly.

2. Install GraalVM native image as follows:

a) Download GraalVM from https://github.com/graalvm/graalvm-ce-builds/releases

b) Unpack and add the `bin` directory to your PATH.

c) Run:

    $ gu install native-image

3. Finally, you can build the native `minirandr` executable using:

    $ clj -T:build native-image

This will create the standalone executable `minirandr` the `target` directory.

## Usage

### Basic usage

First, call `minirandr -l` to list connected screens. Calling `minirandr` without arguments does the same thing. Each screen is listed alongside a numeric index (0, 1, 2 etc...) and its native resolution. For example:

    $ minirandr
    0: eDP1 1920x1080
    1: DP1 1920x1200
    2: HDMI2 2560x1080

Then decide which screens you want to activate by supplying their screen indexes to `minirandr`. The screens will be arranged *left to right* in the order that the indexes are supplied. For example, to place screen 1 at the far left, then 2 in the middle, and 0 at the right, run the following:

    $ minirandr 1 2 0
    xrandr --output DP1 --auto --output HDMI2 --auto --right-of DP1 --output eDP1 --auto --right-of HDMI2

As you can see, the underlying `xrandr` command being run is conveniently shown.

The `xrandr` command also has the concept of a "primary" screen which commonly determines where your window manager panel appears or where new windows appear by default. The primary screen can be specified by adding the letter "p" to the screen index. In our example setup, to set screen 2 as primary, run the following:

    $ minirandr 1 2p 0
    xrandr --output DP1 --auto --output HDMI2 --auto --primary --right-of DP1 --output eDP1 --auto --right-of HDMI2

Any screen indexes missing from the command will switch that screen off. In our example setup, to switch off screen 0 and leave only screens 1 and 2 active, run the following:

    $ minirandr 1 2p
    xrandr --output DP1 --auto --output HDMI2 --auto --primary --right-of DP1 --output eDP1 --off

### Single screen

To configure a screen as the single screen. Run:

    $ minirandr -s 1p
    xrandr --output DP1 --auto --primary ---output eDP1 --off --output HDMI2 -off

You can also quickly configure screen 0 as the single primary screen. This is especially useful when you know you only have a single screen, for example, after disconnecting your laptop from all external screens, run:

    $ minirandr -s
    xrandr --output eDP1 --auto --primary ---output DP1 --off --output HDMI2 -off

### Mirror screens

To mirror screens, for example screens 0 and 1 with 1 as primary while maintaining their native resolution and setting, run:

    $ minirandr -m 0 2p
    xrandr --output eDP1 --auto --output HDMI2 --auto --primary --same-as eDP1 --output DP1 --off

And as a quick shortcut, to mirror all screens while setting screen 0 as primary, run:

    $ minirandr -m
    xrandr --output eDP1 --auto --primary --output DP1 --auto --same-as eDP1 --output HDMI2 --auto --same-as DP1

### Help

Finally, to see the help, run:

    $ minirandr -h

That's it!


CHANGELOG.md
# Changelog

## 0.2.0

- Bump version of `clojure` to 1.11.1.
- When setting single screen, can now specify which screen to make the single screen.
- Can now mirror screens.
- Add changelog.

## 0.1.1

- Initial release.


test/com/safehammad/minirandr_test.clj
(ns com.safehammad.minirandr-test
  (:require [clojure.test :refer [deftest is are testing]]
            [com.safehammad.minirandr :as minirandr]
            [clojure.string :as str]))

(def sample-xrandr-output
    "eDP1 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 310mm x 170mm
    1920x1080     60.00*+  59.93
    1680x1050     59.88
    1400x1050     59.98
    1600x900      60.00    59.95    59.82
    1280x1024     60.02
    1400x900      59.96    59.88
    1280x960      60.00
    1368x768      60.00    59.88    59.85
    1280x800      59.81    59.91
    1280x720      59.86    60.00    59.74
    1024x768      60.00
    1024x576      60.00    59.90    59.82
    960x540       60.00    59.63    59.82
    800x600       60.32    56.25
    864x486       60.00    59.92    59.57
    640x480       59.94
    720x405       59.51    60.00    58.99
    640x360       59.84    59.32    60.00
    DP1 disconnected (normal left inverted right x axis y axis)
    DP2 disconnected (normal left inverted right x axis y axis)
    HDMI1 disconnected (normal left inverted right x axis y axis)
    HDMI2 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 530mm x 300mm
    1920x1080     60.00*+  74.97    50.00    59.94
    1920x1080i    60.00    50.00    59.94
    1680x1050     59.88
    1280x1024     75.02    60.02
    1440x900      59.90
    1280x960      60.00
    1280x720      60.00    50.00    59.94
    1024x768      75.03    70.07    60.00
    832x624       74.55
    800x600       72.19    75.00    60.32    56.25
    720x576       50.00
    720x480       60.00    59.94
    640x480       75.00    72.81    66.67    60.00    59.94
    720x400       70.08
    VIRTUAL1 disconnected (normal left inverted right x axis y axis)")

(def sample-screen-map {0 {:screen "eDP1", :connected true, :resolution "1920x1080"},
                        1 {:screen "HDMI2", :connected true, :resolution "1920x1080"},
                        2 {:screen "DP1", :connected false, :resolution nil},
                        3 {:screen "DP2", :connected false, :resolution nil},
                        4 {:screen "HDMI1", :connected false, :resolution nil},
                        5 {:screen "VIRTUAL1", :connected false, :resolution nil}})

(def sample-screen-choice [{:primary false, :screen-index 0, :off false}   ; screen-spec 0 1p
                           {:primary true, :screen-index 1, :off false}])

(def sample-off-screens [{:screen-index 2, :off true, :primary false}  ; screen-spec 0 1p
                         {:screen-index 3, :off true, :primary false}
                         {:screen-index 4, :off true, :primary false}
                         {:screen-index 5, :off true, :primary false}])

(deftest parse-xrandr-test
  (is (= sample-screen-map (minirandr/parse-xrandr sample-xrandr-output))))

(deftest valid-args-test
  (testing "Valid args"
    (are [args] (minirandr/valid-args? sample-screen-map args)
         ["0"]     ; connected index
         ["1"]     ; another connected index
         ["0p"]))  ; set as primary
  (testing "Invalid args"
    (are [args] (not (minirandr/valid-args? sample-screen-map args))
         []        ; args must exist - we won't switch off all screens!
         ["x"]     ; not a screen index
         ["0x"]    ; suffix can only be p
         ["2"]     ; screen not connected
         ["6"])))  ; screen index out of range

(deftest connected-screen-indexes-test
  (is (= #{"0" "1"} (minirandr/connected-screen-indexes sample-screen-map))))  ; Args: 0 1p

(deftest make-screen-choice-test
  (is (= sample-screen-choice (minirandr/make-screen-choice ["0" "1p"]))))

(deftest off-screens-test
  (is (= sample-off-screens (minirandr/make-off-screens sample-screen-map sample-screen-choice))))

(deftest format-xrandr-cmd-test
  (are [mirror-screens? cmd] (= (str/join " " (minirandr/format-xrandr-cmd sample-screen-map sample-screen-choice sample-off-screens mirror-screens?)) cmd)
       false "xrandr --output eDP1 --auto --output HDMI2 --auto --primary --right-of eDP1 --output DP1 --off --output DP2 --off --output HDMI1 --off --output VIRTUAL1 --off"
       true  "xrandr --output eDP1 --auto --output HDMI2 --auto --primary --same-as eDP1 --output DP1 --off --output DP2 --off --output HDMI1 --off --output VIRTUAL1 --off"))


src/com/safehammad/minirandr.clj
(ns com.safehammad.minirandr
  (:require [clojure.string :as str]
            [clojure.set :as set]
            [clojure.java.shell :as shell]
            [clojure.tools.cli :as cli])

  (:gen-class))

(def version-string "0.2.0")

(defn ->screen
  "Create screen entity."
  [[screen connected & [resolution]]]
  (let [connected? (= connected "connected")]
    {:screen screen
     :connected connected?
     :resolution (when connected? resolution)}))

(defn parse-xrandr
  "Parse xrandr output to return a sorted screen map of {screen-name connected resolution, ...}.

  For example: {0 {:screen \"eDP1\", :connected true :resolution \"1920x1080\"}
                1 {:screen \"DP1\", :connected true :resolution \"1920x1200\"}
                2 {:screen \"HDMI1\", :connected false :resolution nil}"
  [xrandr-output]
  (into {} (->> (str/split-lines xrandr-output)
                (map str/trim)
                (partition-all 2 1)
                (filter #(re-matches #".* (connected|disconnected) .*" (first %)))
                (map (partial map #(str/split % #" ")))
                (map (partial mapcat (partial take 2)))
                (map ->screen)
                (sort-by (comp not :connected))  ; group connected items {:connected true} first
                (map-indexed vector))))

(comment
  (parse-xrandr
  "eDP1 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 310mm x 170mm
   1920x1080     60.00*+  59.93
   1680x1050     59.88
   1400x1050     59.98
   1600x900      60.00    59.95    59.82
   1280x1024     60.02
   1400x900      59.96    59.88
   1280x960      60.00
   1368x768      60.00    59.88    59.85
   1280x800      59.81    59.91
   1280x720      59.86    60.00    59.74
   1024x768      60.00
   1024x576      60.00    59.90    59.82
   960x540       60.00    59.63    59.82
   800x600       60.32    56.25
   864x486       60.00    59.92    59.57
   640x480       59.94
   720x405       59.51    60.00    58.99
   640x360       59.84    59.32    60.00
DP1 disconnected (normal left inverted right x axis y axis)
DP2 disconnected (normal left inverted right x axis y axis)
HDMI1 disconnected (normal left inverted right x axis y axis)
HDMI2 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 530mm x 300mm
   1920x1080     60.00*+  74.97    50.00    59.94
   1920x1080i    60.00    50.00    59.94
   1680x1050     59.88
   1280x1024     75.02    60.02
   1440x900      59.90
   1280x960      60.00
   1280x720      60.00    50.00    59.94
   1024x768      75.03    70.07    60.00
   832x624       74.55
   800x600       72.19    75.00    60.32    56.25
   720x576       50.00
   720x480       60.00    59.94
   640x480       75.00    72.81    66.67    60.00    59.94
   720x400       70.08
VIRTUAL1 disconnected (normal left inverted right x axis y axis)"))

(defn print-screens
  "Return string formatted 0-based indexed list of screens to choose from."
  [screen-map]
  (str/join "\n" (for [[i {:keys [screen connected resolution] }] screen-map
                       :when connected]
                   (str i ": " screen " " resolution))))

(defn make-screen-choice
  "Given command line args, return set of screen-indexes chosen in order, together with whether they're primary.
  For example: 2p 0 = ({:screen-index \"2\" :primary true} {:screen-index \"0\" :primary false})"
  [screen-specs]
  (map #(hash-map :screen-index (parse-long (subs % 0 1))
               :primary (str/ends-with? % "p")
               :off false)
       screen-specs))

(defn make-off-screens
  "Given a map of screens and the choice of screens to turn on, return a map of screens to switch off."
  [screen-map screen-choice]
  (let [chosen-screen-indexes (map :screen-index screen-choice)
        all-screen-indexes    (keys screen-map)
        off-screen-indexes    (remove (set chosen-screen-indexes) all-screen-indexes)]
    (map
      #(hash-map :screen-index %
                 :primary false
                 :off true)
      off-screen-indexes)))

(defn extend-command
  "Extend the xrandr command with the given screen config."
  [cmd screen off primary previous-screen mirror-screens?]
  (let [cmd (conj cmd "--output" screen)]
    (if off
      (conj cmd "--off")
      (cond-> (conj cmd "--auto" )
        primary         (conj "--primary")
        previous-screen (conj (if mirror-screens? "--same-as" "--right-of") previous-screen)))))

(defn format-xrandr-cmd
  "Create the xrandr command to execute as a vector of strings."
  [screen-map screen-choice off-screens mirror-screens?]
  (loop [cmd             ["xrandr"]
         screen-config   (concat screen-choice off-screens)
         previous-screen nil]
    (if (seq screen-config)
      (let [{:keys [screen-index primary off]} (first screen-config)
            {:keys [screen]}                   (get screen-map screen-index)
            cmd                                (extend-command cmd screen off primary previous-screen mirror-screens?)]
        (recur cmd (rest screen-config) screen))
      cmd)))

(defn run-xrandr!
  "Format and run the xrandr command."
  [screen-map screen-specs mirror-screens?]
  (let [screen-choice (make-screen-choice screen-specs)
        off-screens   (make-off-screens screen-map screen-choice)
        cmd           (format-xrandr-cmd screen-map screen-choice off-screens mirror-screens?)
        result        (apply shell/sh cmd)]
    (str (str/join " " cmd) "\n" (:err result))))

(defn connected-screen-indexes
  "Return the indexes of connected screens as strings."
  [screen-map]
  (->> screen-map
       (filter (comp :connected val))
       (map (comp str key))))

(defn make-screen-specs
  "Given information provided on command line, create screen specs."
  [single-screen? mirror-screens? screen-map arguments]
  (cond
    single-screen?  (if (seq arguments)
                      (subvec arguments 0 1)
                      ["0p"])
    mirror-screens? (if (seq arguments)
                      arguments
                      (let [[spec & specs] (connected-screen-indexes screen-map)]
                        (apply vector (str spec "p") specs)))
    :else            arguments))

(defn valid-args?
  "Ensure indexes parsed as args correspond to connected screens."
  [screen-map args]
  (let [connected-indexes (set (connected-screen-indexes screen-map))
        parsed-args       (map (comp second (partial re-matches #"(\d+)p?")) args)]
    (and
      (seq parsed-args)
      (every? some? parsed-args)
      (set/subset? (set parsed-args) connected-indexes))))

(defn screen-map!
  "Run xrandr command to return a screen map."
  []
  (-> (shell/sh "xrandr") :out parse-xrandr))

;; Command line entrypoint

(def usage "Usage: minirandr [OPTION] [SCREEN]...")

(def instructions "First list all connected screens by running:

    $ minirandr -l  # This is the same as running minirandr with no arguments

Connected screens will be listed by index detailing their name and resolution, for example:

    0: eDP1 1920x1080
    1: HDMI2 2560x1080

To configure screens, provide screen specs as indexes in the order that they are positioned, left to right. An optional `p` can be added to an index to make a screen primary. For example, to configure screen 1 to be left of screen 0, and for screen 0 to be primary, run:

    $ minirandr 1 0p

To configure screen 2 as the single primary screen, run:

    $ minirandr -s 2p

To quickly configure screen 0 as the single primary screen, run:

    $ minirandr -s

To mirror screens 0 and 1 with 1 as primary while maintaining their native resolution and setting, run:

    $ minirandr -m 0 1p

To mirror all secreen setting screen 0 as primary while maintaining their native resolution and setting, run:

    $ minirandr -m")

(def cli-opts
  [["-l" "--list-screens" "List connected screens"]
   ["-s" "--single-screen" "Configure as only screen; if no screen specs provided, choose first connected screen and set as primary"]
   ["-m" "--mirror-screens" "Mirror screens; if no screen specs provided, mirror all connected screens and set first as primary"]
   ["-v" "--version" "Display version"]
   ["-h" "--help" "Display help"]])

(defn main!
  "Main entry point."
  [args]
  (let [parsed-opts                                (cli/parse-opts args cli-opts)
        {:keys [options arguments errors summary]} parsed-opts
        {:keys [list-screens single-screen
                mirror-screens version help]}      options
        screen-map                                 (screen-map!)
        screen-specs                               (make-screen-specs single-screen mirror-screens screen-map arguments)
        list-screens                               (or list-screens (every? empty? [options screen-specs]))]
    (cond
      errors                                      (str/join "\n" errors)
      help                                        (str/join "\n\n" [usage summary instructions])
      version                                     (str "minirandr v" version-string)
      list-screens                                (print-screens screen-map)
      (not (valid-args? screen-map screen-specs)) (str "Invalid screen specs. "
                                                    "The following screens are available:\n\n"
                                                    (print-screens screen-map)
                                                    "\n\nPlease run `minirandr --help` for more information.")
      :else                                       (run-xrandr! screen-map screen-specs mirror-screens))))

(defn -main
  [& args]
  (println (main! args))
  (System/exit 0))
